File: pages/login/login_page.py
import re
from pages.__base import BasePage
from elements.__login import *
from playwright.async_api import Page, expect


class LoginPage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page)
        self.action = LogInPageInteraction(page)
        self.check = LoginPageValidation(page)


class LogInPageInteraction(BasePage):
    async def open_page(self):
        await self.page.goto(Url.login)
        assert r'/login' in self.page.url, f"Incorrect URL: {self.page.url} detected"

    async def username_insert(self, username: str):
        await expect(self._find(Interactor.username_input)).to_be_empty()
        await self._type(Interactor.username_input, username)
        await expect(self._find(Interactor.username_input)).not_to_be_empty()

    async def password_insert(self, password: str):
        await expect(self._find(Interactor.password_input)).to_be_empty()
        await self._type(Interactor.password_input, password)
        await expect(self._find(Interactor.password_input)).not_to_be_empty()

    async def click_login_btn(self):
        await self._click(Interactor.login_btn)

    async def close_success_banner(self):
        await expect(self._find(PageInfo.banner_main)).to_contain_text("You logged out of the secure area!")
        await self._click(PageInfo.banner_close)
        await expect(self._find(PageInfo.banner_main)).not_to_be_visible()


class LoginPageValidation(BasePage):
    async def header_presence(self):
        await self._look(PageInfo.header)
        page_header = await self._find(PageInfo.header).text_content()
        assert page_header == "Login Page", f"The Current page title:{page_header} not match with 'Login Page'"

    async def subheader_presence(self):
        await self._look(PageInfo.sub_header)
        await expect(self._find(PageInfo.sub_header)).to_contain_text(
            "This is where you can log into the secure area")
        await expect(self._find(PageInfo.sub_header)).to_contain_text(
            "If the information is wrong you should see error messages")

    async def username_field(self):
        await self._look(Interactor.username_label)
        await expect(self._find(Interactor.username_label)).to_have_text("Username")
        
        await self._look(Interactor.username_input)
        await expect(self._find(Interactor.username_input)).to_be_enabled()
        await expect(self._find(Interactor.username_input)).to_have_value('')

    async def password_field(self):
        await self._look(Interactor.password_label)
        await expect(self._find(Interactor.password_label)).to_have_text("Password")

        await self._look(Interactor.password_input)
        await expect(self._find(Interactor.password_input)).to_be_enabled()
        await expect(self._find(Interactor.password_input)).to_have_value('')

    async def login_button(self):
        await self._look(Interactor.login_btn)
        await expect(self._find(Interactor.login_btn)).to_be_enabled()
        await expect(self._find(Interactor.login_btn)).to_contain_text("Login")

    async def logout_banner_presence(self):
        await self._look(PageInfo.banner_main)
        await expect(self._find(PageInfo.banner_main)).to_contain_text("You logged out of the secure area!")

    async def logout_banner_close(self):
        await self._look(PageInfo.banner_close)
        await expect(self._find(PageInfo.banner_close)).to_be_enabled()

File: pages/login/loginPage.ts
import { BasePage } from '../__base';
import { Interactor, PageInfo, Url } from '../../elements/__login';
import { expect, Page } from '@playwright/test';

export class LoginPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }


  // Login Page Interaction
  async openPage() {
    await this.page.goto(Url.login);
    expect(this.page.url).toContain('/login');
  }

  async usernameInsert(username: string) {
    expect(await this._find(Interactor.username_input).isEditable()).toBeTruthy();
    await this._type(Interactor.username_input, username);
    await expect(this._find(Interactor.username_input)).toBeEditable();
    await this._capture('Username Insert');
  }

  async passwordInsert(password: string) {
    expect(await this._find(Interactor.password_input).isVisible()).toBeTruthy();
    await this._type(Interactor.password_input, password);
    await this._capture('Password Insert');
  }

  async clickLoginBtn() {
    await this._click(Interactor.login_btn);
  }

  async closeSuccessBanner() {
    await expect(this._find(PageInfo.banner_main)).toContainText('You logged out of the secure area!');
    await this._click(PageInfo.banner_close);
    await expect(this._find(PageInfo.banner_main)).not.toBeVisible();
  }

  async closeInvalidUserBanner() {
    await expect(this._find(PageInfo.banner_main)).toContainText('username is invalid!');
    await this._click(PageInfo.banner_close);
    await expect(this._find(PageInfo.banner_main)).not.toBeVisible();
  }

  async closeInvalidPasswordBanner() {
    await expect(this._find(PageInfo.banner_main)).toContainText('password is invalid!');
    await this._click(PageInfo.banner_close);
    await expect(this._find(PageInfo.banner_main)).not.toBeVisible();
  }

  // Login Page Validation
  async urlRedirection() {
    expect(this.page.url).toContainEqual('/login');
  }

  async HeaderPresence() {
    await this._look(PageInfo.header)
    const page_header = await this._find(PageInfo.header).textContent();
    expect(page_header).toBe("Login Page");
  }

  async SubheaderPresence() {
    await this._look(PageInfo.sub_header)
    await expect(this._find(PageInfo.sub_header)).toContainText(
      "This is where you can log into the secure area");
    await expect(this._find(PageInfo.sub_header)).toContainText(
      "If the information is wrong you should see error messages");
  }

  async usernameFieldPresence() {
    await this._look(Interactor.username_label)
    await expect(this._find(Interactor.username_label)).toHaveText("Username")

    await this._look(Interactor.username_input)
    await expect(this._find(Interactor.username_input)).toBeEnabled()
    await expect(this._find(Interactor.username_input)).toBeEmpty()
  }

  async passwordFieldPresence() {
    await this._look(Interactor.password_label)
    await expect(this._find(Interactor.password_label)).toHaveText("Password")

    await this._look(Interactor.password_input)
    await expect(this._find(Interactor.password_input)).toBeEnabled()
    await expect(this._find(Interactor.password_input)).toBeEmpty()
  }

  //Login Page Validation: Successful Banner
  async loginButtonPresence() {
    await this._look(Interactor.login_btn)
    await expect(this._find(Interactor.login_btn)).toBeEnabled()
    await expect(this._find(Interactor.login_btn)).toContainText("Login")
  }

  async logoutBannerPresence() {
    await this._look(PageInfo.banner_main)
    await expect(this._find(PageInfo.banner_main)).toContainText("You logged out of the secure area!")
    await this._capture("Logout Success Banner")
  }

  async logoutCloseBannerPresence() {
    await this._look(PageInfo.banner_close)
    await expect(this._find(PageInfo.banner_close)).toBeEnabled()
  }

  // Login Page Validation: Invalid Banner
  async invalidBannerUsernamePresence() {
    await this._look(PageInfo.banner_main)
    await expect(this._find(PageInfo.banner_main)).toContainText("username is invalid!")
    await expect(this._find(PageInfo.banner_close)).toBeEnabled()
    await this._capture("Invalid Username Banner")
  }

  async invalidBannerPasswordPresence() {
    await this._look(PageInfo.banner_main)
    await expect(this._find(PageInfo.banner_main)).toContainText("password is invalid!")
    await expect(this._find(PageInfo.banner_close)).toBeEnabled()
    await this._capture("Invalid Password Banner")
  }
}
